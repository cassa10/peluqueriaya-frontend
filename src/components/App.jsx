import React from 'react';
import {Route, Switch} from 'react-router';
import {BrowserRouter} from 'react-router-dom';
import Container from "@material-ui/core/Container";
import Bar from "./Bar";
import LoggedOutHomePage from "../views/LoggedOutHomePage";
import SearchResults from "../views/SearchResults";

const App = () => {

    return (
        <BrowserRouter>
            <Bar/>
            <Container maxWidth="lg" >
                <Switch>
                  <Route exact path="/" render={ (props) => LoggedOutHomePage({...props})}/>
                  <Route path="/search" render={ (props) => SearchResults({...props})} />
                </Switch>
            </Container>
        </BrowserRouter>
    );
};

export default App;