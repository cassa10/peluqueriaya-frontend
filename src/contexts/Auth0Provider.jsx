import React, { createContext, useContext, useEffect, useState } from "react";
import createAuth0Client from "@auth0/auth0-spa-js";
import PropTypes from "prop-types";
import { URI_CASA } from "../utils/constants";

export const Auth0Context = createContext();
export const useAuth0 = () => useContext(Auth0Context);

const Auth0Provider = ({ history, children, ...initOptions }) => {
  const [email, setEmail] = useState();
  const [auth0Client, setAuth0] = useState();
  const [loading, setLoading] = useState(true);
  const [isAuthenticated, setIsAuthenticated] = useState(false);

  const onRedirectCallback = (appState) => {
    history.push(
      appState && appState.targetUrl
        ? appState.targetUrl
        : window.location.pathname
    );
  };

  useEffect(() => {
    const initAuth0 = async () => {
      const auth0FromHook = await createAuth0Client(initOptions);
      setAuth0(auth0FromHook);
      if (
        window.location.search.includes("code=") &&
        window.location.search.includes("state=")
      ) {
        const { appState } = await auth0FromHook.handleRedirectCallback();
        onRedirectCallback(appState);
      }
      const isAuthenticated = await auth0FromHook.isAuthenticated();
      setIsAuthenticated(isAuthenticated);
      const { email = null } =
        isAuthenticated && (await auth0FromHook.getUser());
      setEmail(email);
      setLoading(false);
    };
    initAuth0();
    // eslint-disable-next-line
  }, []);

  const login = (redirect_uri) => {
    if (isAuthenticated) {
      history.push(redirect_uri);
    } else {
      auth0Client.loginWithRedirect({ appState: { targetUrl: redirect_uri } });
    }
  };

  const logout = () => auth0Client.logout({ returnTo: URI_CASA });

  return (
    <Auth0Context.Provider
      value={{
        email,
        loading,
        isAuthenticated,
        login,
        logout,
        getTokenSilently: (...p) => auth0Client.getTokenSilently(...p),
      }}
    >
      {children}
    </Auth0Context.Provider>
  );
};

Auth0Provider.propTypes = {
  history: PropTypes.object.isRequired,
  children: PropTypes.node.isRequired,
};

export default Auth0Provider;
