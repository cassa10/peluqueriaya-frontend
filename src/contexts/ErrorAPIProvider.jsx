import React, { createContext, useContext, useEffect, useState } from "react";
import PropTypes from "prop-types";
import get from "lodash/get";
import { useHistory } from "react-router-dom";
import PaginaError from "../views/PaginaError";

const ErrorAPIContext = createContext();
export const useError = () => useContext(ErrorAPIContext);

const ErrorAPIProvider = ({ children }) => {
  const history = useHistory();
  const [error, setError] = useState();

  useEffect(() => {
    return history.listen(() => setError(undefined));
    // eslint-disable-next-line
  }, []);

  const mostrarContenido = () => {
    if (error) {
      const { status, ...paginaErrorProps } = get(error, "response.data", {});
      return <PaginaError {...paginaErrorProps} />;
    } else {
      return children;
    }
  };

  return (
    <ErrorAPIContext.Provider value={{ setError }}>
      {mostrarContenido()}
    </ErrorAPIContext.Provider>
  );
};

ErrorAPIProvider.propTypes = {
  children: PropTypes.element.isRequired,
};

export default ErrorAPIProvider;
