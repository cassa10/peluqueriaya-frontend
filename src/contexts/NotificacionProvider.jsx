import React, {createContext, useContext, useEffect, useMemo, useState} from 'react';
import Snackbar from "@material-ui/core/Snackbar";
import Alert from "@material-ui/lab/Alert";
import PropTypes from "prop-types";
import isEqual from "lodash/isEqual";

const NotificacionContext = createContext();
export const useNotificacion = () => useContext(NotificacionContext);

const notificacionInicial = {
    mensaje: "",
    severidad: "info"
};

const NotificacionProvider = ({children}) => {
    const [abrir, setAbrir] = useState(false);
    const [notificacion, setNotificacion] = useState(notificacionInicial);

    useEffect(() => {
        if(!isEqual(notificacion, notificacionInicial)) {
            setAbrir(true);
        }
    }, [notificacion]);

    const contextPayload = useMemo(
        () => ({setNotificacion}),
        [setNotificacion]
    );

    const handleClose = (event, reason) => {
        if (reason !== "clickaway") {
            setAbrir(false);
            setNotificacion(notificacionInicial);
        }
    };

    return <NotificacionContext.Provider value={contextPayload}>
        {children}
        <Snackbar open={abrir} autoHideDuration={10000} onClose={handleClose}>
            <Alert elevation={6} variant="filled" onClose={handleClose} severity={notificacion.severidad}>
                {notificacion.mensaje}
            </Alert>
        </Snackbar>
    </NotificacionContext.Provider>;


};

NotificacionProvider.propTypes = {
    children: PropTypes.element.isRequired
};

export default NotificacionProvider;