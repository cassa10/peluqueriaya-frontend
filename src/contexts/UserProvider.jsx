import React, {useState, useEffect, useContext, createContext} from "react";
import createAuth0Client from "@auth0/auth0-spa-js";
import PropTypes from "prop-types";
import axios from "axios";
import {
    CLIENTE,
    PELUQUERO,
    VISITANTE,
    REGISTRADO,
    URI_CASA,
    URI_LOGIN_CLIENTE,
    URI_LOGIN_PELUQUERO, PENDIENTE
} from "../assets/constants";

export const UserContext = createContext();
export const useUser = () => useContext(UserContext);

const rolesIniciales = {
    [CLIENTE]: VISITANTE,
    [PELUQUERO]: VISITANTE
};

const UserProvider = ({history, children, ...initOptions}) => {
    const [user, setUser] = useState();
    const [auth0Client, setAuth0] = useState();
    const [loading, setLoading] = useState(true);
    const [roles, setRoles] = useState(rolesIniciales);

    const onRedirectCallback = appState => {
        history.push(
            appState && appState.targetUrl
                ? appState.targetUrl
                : window.location.pathname
        );
    };

    useEffect(() => {
        const initAuth0 = async () => {
            const auth0FromHook = await createAuth0Client(initOptions);
            setAuth0(auth0FromHook);
            if (window.location.search.includes("code=") &&
                window.location.search.includes("state=")) {
                const {appState} = await auth0FromHook.handleRedirectCallback();
                onRedirectCallback(appState);
            }
            const isAuthenticated = await auth0FromHook.isAuthenticated();
            if (isAuthenticated) {
                const user = await auth0FromHook.getUser();
                setUser(user);
                const token = await auth0FromHook.getTokenSilently();
                const {data} = await axios.get('http://localhost:8080/roles', {
                    headers: {Authorization: `Bearer ${token}`}
                });
                setRoles(data);
            }
            setLoading(false);
        };
        initAuth0();
        // eslint-disable-next-line
    }, []);

    const rolesHandler = {
        [CLIENTE]: {
            uri_login: URI_LOGIN_CLIENTE,
            noEstaRegistradoEnOtroRol: () => roles[PELUQUERO] !== REGISTRADO
        },
        [PELUQUERO]: {
            uri_login: URI_LOGIN_PELUQUERO,
            noEstaRegistradoEnOtroRol: () => roles[CLIENTE] !== REGISTRADO
        }
    };

    const login = (rol) => {
        if (rolesHandler[rol].noEstaRegistradoEnOtroRol()) {
            auth0Client.loginWithRedirect({appState: {targetUrl: rolesHandler[rol].uri_login}});
        } else {
            history.push(rolesHandler[rol].uri_login);
        }
    };

    const logout = () => auth0Client.logout({returnTo: URI_CASA});

    const empezarRegistro = (rol) => {
        if (roles[rol] === VISITANTE) {
            setRoles(prevState => ({...prevState, [rol]: PENDIENTE}));
        }
    };

    const abandonarRegistro = (rol) => {
        if (roles[rol] === PENDIENTE) {
            setRoles(prevState => ({...prevState, [rol]: VISITANTE}));
            if (rolesHandler[rol].noEstaRegistradoEnOtroRol()) logout();
        }
    };

    const registrar = (rol) => setRoles(prevState => ({...prevState, [rol]: REGISTRADO}));

    return (<UserContext.Provider
            value={{user, loading, roles, setRoles, login, logout, empezarRegistro, abandonarRegistro,
                registrar, getTokenSilently: (...p) => auth0Client.getTokenSilently(...p)
            }}>
            {children}
        </UserContext.Provider>
    );
};

UserProvider.propTypes = {
    history: PropTypes.object.isRequired,
    children: PropTypes.node.isRequired
};

export default UserProvider;