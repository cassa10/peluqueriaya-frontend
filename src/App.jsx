import React from 'react';
import styled from 'styled-components';
import CssBaseline from '@material-ui/core/CssBaseline';
import {
    Root,
    getHeader,
    getDrawerSidebar,
    getSidebarContent,
    getContent,
    getFixedScheme
} from '@mui-treasury/layout';
import theme from "./assets/theme";
import ContenidoBarraLateral from "./components/sidebar/ContenidoBarraLateral";
import CollapseBtnStyled from "./components/sidebar/CollapseBtnStyled";
import ContenidoHeader from "./components/ContenidoHeader";
import {useUser} from "./contexts/UserProvider";
import LinearProgress from "@material-ui/core/LinearProgress";
import ContenidoCuerpo from "./components/ContenidoCuerpo";

const Header = getHeader(styled);
const DrawerSidebar = getDrawerSidebar(styled);
const SidebarContent = getSidebarContent(styled);
const Content = getContent(styled);

const scheme = getFixedScheme();

const App = () => {
    const {loading} = useUser();

    return (
        <Root theme={theme} scheme={scheme}>
            {({state: {sidebar}}) => {
                if (loading) {
                    return <LinearProgress color="secondary"/>;
                }
                return <>
                    <CssBaseline/>
                    <Header color="secondary">
                        <ContenidoHeader/>
                    </Header>
                    <DrawerSidebar sidebarId="primarySidebar">
                        <SidebarContent>
                            <ContenidoBarraLateral collapsed={sidebar.primarySidebar.collapsed}/>
                        </SidebarContent>
                        <CollapseBtnStyled/>
                    </DrawerSidebar>
                    <Content>
                        <ContenidoCuerpo/>
                    </Content>
                </>;
            }}
        </Root>
    );
};

export default App;