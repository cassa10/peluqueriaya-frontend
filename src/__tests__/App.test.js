import {shallow} from "enzyme";
import App from "../components/App";
import React from "react";
import * as userProvider from "../contexts/UserProvider";

describe('App', () => {
    it('App levanta sin problemas', () => {
        userProvider.useUser = jest.fn().mockReturnValue(true);
        const component = shallow(<App />);
        expect(component).toMatchSnapshot();
    });
});