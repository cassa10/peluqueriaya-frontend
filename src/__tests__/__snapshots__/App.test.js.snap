// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App levanta en loading 1`] = `
<NotificacionProvider>
  <div
    className="makeStyles-root-1"
  >
    <Header
      color="primary"
    >
      <ContenidoHeader />
    </Header>
    <Component>
      <DrawerSidebar
        sidebarId="primarySidebar"
      >
        <SidebarContent>
          <ContenidoBarraLateral />
        </SidebarContent>
        <CollapseBtnStyled />
      </DrawerSidebar>
    </Component>
    <Content
      className="makeStyles-content-2"
    >
      <ErrorAPIProvider>
        <ContenidoCuerpo />
      </ErrorAPIProvider>
    </Content>
    <div
      className="makeStyles-footer-3"
    >
      <Footer>
        <WithStyles(ForwardRef(Typography))
          align="center"
          variant="caption"
        >
          © 2020 - MIT License
        </WithStyles(ForwardRef(Typography))>
      </Footer>
    </div>
  </div>
</NotificacionProvider>
`;

exports[`App App levanta sin problemas 1`] = `
<NotificacionProvider>
  <div
    className="makeStyles-root-1"
  >
    <Header
      color="primary"
    >
      <ContenidoHeader />
    </Header>
    <Component>
      <DrawerSidebar
        sidebarId="primarySidebar"
      >
        <SidebarContent>
          <ContenidoBarraLateral />
        </SidebarContent>
        <CollapseBtnStyled />
      </DrawerSidebar>
    </Component>
    <Content
      className="makeStyles-content-2"
    >
      <ErrorAPIProvider>
        <ContenidoCuerpo />
      </ErrorAPIProvider>
    </Content>
    <div
      className="makeStyles-footer-3"
    >
      <Footer>
        <WithStyles(ForwardRef(Typography))
          align="center"
          variant="caption"
        >
          © 2020 - MIT License
        </WithStyles(ForwardRef(Typography))>
      </Footer>
    </div>
  </div>
</NotificacionProvider>
`;
